### PR TITLE
Bug fix and add documentation for metric_constraints

### PR DESCRIPTION
**Pull request by qingyun-wu**
- Fix a bug in adding `time_total_s`
- Add a notebook example and documentation for the `metric_constraints` parameter in `AutoML.fit()`
- Update a test in `test/automl/test_notebook_example.py` to verify the good performance of flaml

### CODE REVIEW

File: notebook/automl_classification.ipynb [hunk 1]
```diff
@@ -562,9 +1268,9 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "accuracy = 0.6720332824110467\n",
-      "roc_auc = 0.7253276908529442\n",
-      "log_loss = 0.6034449031876942\n"
+      "accuracy = 0.6533527134657313\n",
```
Comment (sonichi) [inline notebook/automl_classification.ipynb:1271]: what causes the accuracy drop here?

File: notebook/automl_classification.ipynb [hunk 2]
```diff
@@ -1258,15 +2210,1234 @@
...
+    "automl = AutoML()\n",
+    "settings = {\n",
+    "    \"time_budget\": 30,  # total running time in seconds\n",
+    "    \"metric\": custom_metric,  # pass the custom metric funtion here\n",
+    "    \"task\": 'classification',  # task type\n",
+    "    \"log_file_name\": 'airlines_experiment_custom_metric_constrained.log',  # flaml log file\n",
+    "    \"metric_constraints\": metric_constraints,\n",
+    "}\n",
+    "\n",
+    "automl.fit(X_train = X_train, y_train = y_train, **settings)"
```
Comment (sonichi) [inline notebook/automl_classification.ipynb:3429]:
```suggestion
    "automl.fit(X_train=X_train, y_train=y_train, **settings)"
```

**Review by sonichi — COMMENTED** [review]
Need to figure out why the accuracy drops.

**Review by sonichi — COMMENTED** [review]
Please fix build error.

File: test/automl/test_notebook_example.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+import sys
+from sklearn.metrics import accuracy_score
```
Comment (sonichi) [inline test/automl/test_notebook_example.py:2]: where is this used?

File: notebook/automl_classification.ipynb [hunk 4]
```diff
@@ -219,12 +111,13 @@
...
    "source": [
     "settings = {\n",
-    "    \"time_budget\": 240,  # total running time in seconds\n",
+    "    \"time_budget\": 120,  # total running time in seconds\n",
     "    \"metric\": 'accuracy',  # can be: 'r2', 'rmse', 'mae', 'mse', 'accuracy', 'roc_auc', 'roc_auc_ovr',\n",
     "                           # 'roc_auc_ovo', 'log_loss', 'mape', 'f1', 'ap', 'ndcg', 'micro_f1', 'macro_f1'\n",
     "    \"task\": 'classification',  # task type\n",
     "    \"log_file_name\": 'airlines_experiment.log',  # flaml log file\n",
     "    \"seed\": 7654321,    # random seed\n",
+    "    \"n_jobs\": 1,\n",
```
Comment (sonichi) [inline notebook/automl_classification.ipynb:120]: remove this? no need to restrict n_jobs.

File: notebook/automl_classification.ipynb [hunk 5]
```diff
@@ -788,14 +710,14 @@
...
-      "flaml (4min) accuracy = 0.6720332824110467\n"
+      "flaml (1 min) accuracy = 0.670713258087003\n"
      ]
     }
    ],
    "source": [
     "print('default xgboost accuracy', '=', 1 - sklearn_metric_loss_score('accuracy', y_pred_xgb, y_test))\n",
     "print('default lgbm accuracy', '=', 1 - sklearn_metric_loss_score('accuracy', y_pred_lgbm, y_test))\n",
-    "print('flaml (4min) accuracy', '=', 1 - sklearn_metric_loss_score('accuracy', y_pred, y_test))"
+    "print('flaml (1 min) accuracy', '=', 1 - sklearn_metric_loss_score('accuracy', y_pred, y_test))"
```
Comment (sonichi) [inline notebook/automl_classification.ipynb:720]:
```suggestion
    "print('flaml (2 min) accuracy', '=', 1 - sklearn_metric_loss_score('accuracy', y_pred, y_test))"
```

File: notebook/automl_classification.ipynb [hunk 6]
```diff
@@ -1256,16 +1203,991 @@
...
+      "               learning_rate=0.681085689470444, max_bin=7, min_child_samples=54,\n",
+      "               n_estimators=6, num_leaves=6, reg_alpha=0.003095681942146277,\n",
+      "               reg_lambda=125.28594199895093, verbose=-1)\n",
+      "[flaml.automl: 03-24 21:01:09] {2283} INFO - fit succeeded\n",
+      "[flaml.automl: 03-24 21:01:09] {2284} INFO - Time taken to find the best model: 1.6895849704742432\n"
+     ]
+    }
+   ],
+   "source": [
+    "metric_constraints = [(\"train_loss\", \"<=\", 0.3), (\"val_loss\", \"<=\",0.3)]\n",
```
Comment (sonichi) [inline notebook/automl_classification.ipynb:2169]: could you print the train_loss and val_loss of the best config?

File: test/automl/test_notebook_example.py [hunk 7]
```diff
@@ -28,6 +32,7 @@ def test_automl(budget=5, dataset_format="dataframe", hpo_method=None):
         "log_file_name": "airlines_experiment.log",  # flaml log file
         "seed": 7654321,  # random seed
         "hpo_method": hpo_method,
+        "n_jobs": 1,
```
Comment (sonichi) [inline test/automl/test_notebook_example.py:35]: remove?

File: notebook/automl_classification.ipynb (hunk 6 above)
Comment (sonichi) [reply notebook/automl_classification.ipynb:1182]: remove this section or make sure the constraints are satisfied.

File: flaml/automl.py [hunk 8]
```diff
@@ -658,6 +660,20 @@ def best_loss(self):
...
+
+    @property
+    def metrics_for_best_config(self):
+        """Returns a float of the best loss, and a dictionary of the auxiliary metrics to log
+        associated with the best config. These two objects correspond to the returned
+        objects by the customized metric function for the config with the best loss."""
+        return self._state.best_loss, self._search_states[
+            self._best_estimator
+        ].best_result.get("metric_for_logging")
+
```
Comment (sonichi) [inline flaml/automl.py:676]: better use getattr() like other properties to prevent error.

**Review by sonichi — COMMENTED** [review]
Still not safe: self._best_estimator may not be in self._search_states.
Please check how best_config() is defined.
And, merge with main.

**Review by sonichi — APPROVED** [review]
